File: backend/app/Models/DonationGoal.php
<?php

namespace App\Models;

use Illuminate\Database\Eloquent\Factories\HasFactory;
use Illuminate\Database\Eloquent\Model;
use Illuminate\Database\Eloquent\Relations\HasMany;
use Illuminate\Database\Eloquent\SoftDeletes;

class DonationGoal extends Model
{
    use HasFactory, SoftDeletes;

    protected $fillable = [
        'title',
        'slug',
        'description',
        'full_details',
        'category',
        'target_amount',
        'current_amount',
        'donors_count',
        'min_donation',
        'priority',
        'start_date',
        'end_date',
        'status',
        'is_featured',
        'featured_image',
        'gallery_images',
        'milestone_updates',
        'impact_metrics',
        'thank_you_message',
        'show_donors',
        'allow_anonymous',
        'rewards',
    ];

    protected $casts = [
        'target_amount' => 'decimal:2',
        'current_amount' => 'decimal:2',
        'min_donation' => 'decimal:2',
        'start_date' => 'datetime',
        'end_date' => 'datetime',
        'is_featured' => 'boolean',
        'show_donors' => 'boolean',
        'allow_anonymous' => 'boolean',
        'gallery_images' => 'array',
        'milestone_updates' => 'array',
        'rewards' => 'array',
    ];

    // Relations
    public function donations(): HasMany
    {
        return $this->hasMany(Donation::class, 'goal_id');
    }

    public function milestones(): HasMany
    {
        return $this->hasMany(DonationGoalMilestone::class, 'goal_id');
    }

    // Scopes
    public function scopeActive($query)
    {
        return $query->where('status', 'active')
            ->where('start_date', '<=', now())
            ->where(function ($q) {
                $q->whereNull('end_date')
                  ->orWhere('end_date', '>', now());
            });
    }

    public function scopeFeatured($query)
    {
        return $query->where('is_featured', true);
    }

    public function scopeByCategory($query, string $category)
    {
        return $query->where('category', $category);
    }

    public function scopeByPriority($query, string $priority)
    {
        return $query->where('priority', $priority);
    }

    public function scopeUrgent($query)
    {
        return $query->where('priority', 'urgent');
    }

    // Accessors
    public function getProgressPercentageAttribute(): float
    {
        if ($this->target_amount == 0) {
            return 0;
        }

        return min(round(($this->current_amount / $this->target_amount) * 100, 2), 100);
    }

    public function getRemainingAmountAttribute(): float
    {
        return max($this->target_amount - $this->current_amount, 0);
    }

    public function getIsCompletedAttribute(): bool
    {
        return $this->current_amount >= $this->target_amount;
    }

    public function getDaysRemainingAttribute(): ?int
    {
        if (!$this->end_date) {
            return null;
        }

        $days = $this->end_date->diffInDays(now(), false);
        return $days < 0 ? abs($days) : 0;
    }

    // Methods
    public function addDonation(float $amount, User $user): void
    {
        $this->increment('current_amount', $amount);
        $this->increment('donors_count');

        // Check and update milestones
        $this->checkMilestones();

        // Mark as completed if target reached
        if ($this->is_completed && $this->status === 'active') {
            $this->update(['status' => 'completed']);
        }
    }

    public function checkMilestones(): void
    {
        $percentage = $this->progress_percentage;

        $this->milestones()
            ->where('is_achieved', false)
            ->where('percentage', '<=', $percentage)
            ->update([
                'is_achieved' => true,
                'achieved_at' => now(),
            ]);
    }
}

File: mobile/lib/models/donation_goal.dart
class DonationGoal {
  final int id;
  final String title;
  final String slug;
  final String description;
  final String? fullDetails;
  final String category;
  final double targetAmount;
  final double currentAmount;
  final double remainingAmount;
  final double progressPercentage;
  final int donorsCount;
  final double minDonation;
  final String priority;
  final DateTime startDate;
  final DateTime? endDate;
  final int? daysRemaining;
  final String status;
  final bool isCompleted;
  final bool isFeatured;
  final String? featuredImage;
  final List<String>? galleryImages;
  final List<dynamic>? milestoneUpdates;
  final String? impactMetrics;
  final String? thankYouMessage;
  final List<DonationGoalMilestone>? milestones;

  DonationGoal({
    required this.id,
    required this.title,
    required this.slug,
    required this.description,
    this.fullDetails,
    required this.category,
    required this.targetAmount,
    required this.currentAmount,
    required this.remainingAmount,
    required this.progressPercentage,
    required this.donorsCount,
    required this.minDonation,
    required this.priority,
    required this.startDate,
    this.endDate,
    this.daysRemaining,
    required this.status,
    required this.isCompleted,
    required this.isFeatured,
    this.featuredImage,
    this.galleryImages,
    this.milestoneUpdates,
    this.impactMetrics,
    this.thankYouMessage,
    this.milestones,
  });

  factory DonationGoal.fromJson(Map<String, dynamic> json) {
    return DonationGoal(
      id: json['id'],
      title: json['title'],
      slug: json['slug'],
      description: json['description'],
      fullDetails: json['full_details'],
      category: json['category'],
      targetAmount: double.parse(json['target_amount'].toString()),
      currentAmount: double.parse(json['current_amount'].toString()),
      remainingAmount: double.parse(json['remaining_amount'].toString()),
      progressPercentage: double.parse(json['progress_percentage'].toString()),
      donorsCount: json['donors_count'],
      minDonation: double.parse(json['min_donation'].toString()),
      priority: json['priority'],
      startDate: DateTime.parse(json['start_date']),
      endDate: json['end_date'] != null
          ? DateTime.parse(json['end_date'])
          : null,
      daysRemaining: json['days_remaining'],
      status: json['status'],
      isCompleted: json['is_completed'],
      isFeatured: json['is_featured'],
      featuredImage: json['featured_image'],
      galleryImages: json['gallery_images'] != null
          ? List<String>.from(json['gallery_images'])
          : null,
      milestoneUpdates: json['milestone_updates'],
      impactMetrics: json['impact_metrics'],
      thankYouMessage: json['thank_you_message'],
      milestones: json['milestones'] != null
          ? (json['milestones'] as List)
              .map((m) => DonationGoalMilestone.fromJson(m))
              .toList()
          : null,
    );
  }

  String get categoryDisplay {
    switch (category) {
      case 'litigation': return 'Paiement de Litiges';
      case 'player_transfer': return 'Achat de Joueurs';
      case 'stadium_renovation': return 'Rénovation du Stade';
      case 'youth_academy': return 'Académie des Jeunes';
      case 'equipment': return 'Équipements';
      case 'debt_payment': return 'Remboursement de Dettes';
      default: return 'Autre';
    }
  }

  String get priorityDisplay {
    switch (priority) {
      case 'urgent': return 'Urgent';
      case 'high': return 'Élevé';
      case 'medium': return 'Moyen';
      case 'low': return 'Faible';
      default: return priority;
    }
  }
}

class DonationGoalMilestone {
  final int id;
  final String title;
  final String? description;
  final double targetAmount;
  final int percentage;
  final bool isAchieved;
  final DateTime? achievedAt;

  DonationGoalMilestone({
    required this.id,
    required this.title,
    this.description,
    required this.targetAmount,
    required this.percentage,
    required this.isAchieved,
    this.achievedAt,
  });

  factory DonationGoalMilestone.fromJson(Map<String, dynamic> json) {
    return DonationGoalMilestone(
      id: json['id'],
      title: json['title'],
      description: json['description'],
      targetAmount: double.parse(json['target_amount'].toString()),
      percentage: json['percentage'],
      isAchieved: json['is_achieved'],
      achievedAt: json['achieved_at'] != null
          ? DateTime.parse(json['achieved_at'])
          : null,
    );
  }
}
